packages/vite/src/node/server/sourcemap.ts

```diff
@@ -5,16 +5,26 @@ export async function injectSourcesContent(
   map: { sources: string[]; sourcesContent?: string[]; sourceRoot?: string },
   file: string
 ) {
-  const sourceRoot = await fs.realpath(
-    path.resolve(path.dirname(file), map.sourceRoot || '')
-  )
+  try {
+    var sourceRoot = await fs.realpath(
+      path.resolve(path.dirname(file), map.sourceRoot || '')
+    )
+  } catch (e) {
+    if (e.code == 'ENOENT') return
```

Comment:
I don't think we need to worry about differentiating between virtual modules and missing files here. We can just assume the `file` argument represents a virtual module if `fs.realpath` fails. So swallowing the `ENOENT` error is enough here.

Reply:
One implication of just ignoring the error here: Virtual modules need to add `sourcesContent` to their sourcemaps manually. If we want to avoid that requirement, we can assume `sources` in virtual modules' sourcemaps contains absolute file paths, or at least resolve them relative to the project root.